lib/LongVectorLoweringPass.cpp

```diff
@@ -720,6 +749,62 @@ Value *LongVectorLoweringPass::visitLoadInst(LoadInst &I) {
   return V;
 }
 
+Value *LongVectorLoweringPass::visitPHINode(PHINode &I) {
+  // TODO Handle PHIs.
+  //
+  // PHIs are tricky because they require their incoming values
+  // to be handled first, which may not have been defined yet.
+  // We can't explicitly visit them because a PHI may depend on itself,
+  // leading to infinite loops. Defer until we have a test case.
+  //
+  // TODO Add PHI instruction with fast math flag to fastmathflags.ll test.
+  llvm_unreachable("PHI node not yet supported");
+}
+
+Value *LongVectorLoweringPass::visitSelectInst(SelectInst &I) {
+  auto *EquivalentCondition = visitOrSelf(I.getCondition());
+  auto *EquivalentTrueValue = visitOrSelf(I.getTrueValue());
+  auto *EquivalentFalseValue = visitOrSelf(I.getFalseValue());
+
+  assert(((EquivalentCondition != I.getCondition()) ||
+          (EquivalentTrueValue != I.getTrueValue()) ||
+          (EquivalentFalseValue != I.getFalseValue())) &&
+         "nothing to lower");
+
+  // We have two cases to handle here:
+  // - when the condition is a scalar to select one of the two long-vector
+  //   alternatives. In this case, we create a single select instruction.
+  // - when the condition is a long-vector, too. In this case, we do an
+  //   element-wise select and construct an aggregate for the result.
+  Value *V = nullptr;
+  if (EquivalentCondition->getType()->isSingleValueType()) {
+    IRBuilder<> B(&I);
+    V = B.CreateSelect(EquivalentCondition, EquivalentTrueValue,
```

Comment:
This is slightly problematic at present. The SPIRV producer does not handle aggregate selections correctly. SPIR-V before 1.4 cannot select between aggregates; however, they are not usually generated in normal flows. There are two options I see:
1. Change this code to always scalarize
2. Fix the spirv producer

Fixing the producer is more general, but also more complicated as it would have to recursively scalarize the selection.

Reply:
I see. I've opted for the first option for now as it was much simpler for me to implement.